File: 5-Problem-A-Day/86. Partition List.java
/**
 * Definition for singly-linked list. public class ListNode { int val; ListNode next; ListNode(int
 * x) { val = x; } }
 */
class Solution {
    public ListNode partition(ListNode head, int x) {
        ListNode first, second, firstPrev, secondPrev;
        first = second = firstPrev = secondPrev = null;
        while (head != null) {
            if (head.val < x) {
                if (first == null)
                    first = head;
                else
                    firstPrev.next = head;
                firstPrev = head;
            } else {
                if (second == null)
                    second = head;
                else
                    secondPrev.next = head;
                secondPrev = head;
            }
            head = head.next;
        }
        if (first == null)
            return second;
        if (second == null)
            return first;
        secondPrev.next = null;
        firstPrev.next = second;
        return first;
    }
}

File: 2021/86. Partition List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode *dummyLeft = new ListNode(), *dummyRight = new ListNode();
        ListNode *prevLeft = dummyLeft, *prevRight = dummyRight;
        while (head != NULL) {
            if (head->val < x) {
                prevLeft->next = head;
                prevLeft = head;                
            } else {
                prevRight->next = head;
                prevRight = head;
            }
            head = head->next;
        }
        prevRight->next = NULL;
        prevLeft->next = dummyRight->next;
        head = dummyLeft->next;
        delete dummyLeft, dummyRight;
        return head;
    }
};

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode *leftHead = new ListNode(), *rightHead = new ListNode();
        ListNode *leftPrev = leftHead, *rightPrev = rightHead;
        while (head != NULL) {
            if (head->val < x) {
                leftPrev->next = head;
                leftPrev = head;
            } else {
                rightPrev->next = head;
                rightPrev = head;
            }            
            head = head->next;
        }
        leftPrev->next = rightHead->next;
        rightPrev->next = NULL;
        head = leftHead->next;
        delete leftHead, rightHead;
        return head;
    }
};